src/box/txn_limbo.c

```diff
@@ -773,38 +784,287 @@ txn_limbo_wait_empty(struct txn_limbo *limbo, double timeout)
 	return 0;
 }
 
-void
+/**
+ * Wait until all the limbo entries receive an lsn.
+ */
+static int
+txn_limbo_wait_persisted(struct txn_limbo *limbo)
+{
+	if (txn_limbo_is_empty(limbo))
+		return 0;
+
+	while (txn_limbo_last_entry(limbo)->lsn <= 0) {
+		fiber_cond_wait(&limbo->persist_cond);
+		if (fiber_is_cancelled()) {
+			diag_set(FiberIsCancelled);
+			return -1;
+		}
+	}
+	return 0;
+}
+
+/**
+ * Fill the reject reason with request data.
+ * The function is not reenterable, use with care.
+ */
+static char *
+reject_str(const struct synchro_request *req)
+{
+	const char *type_name = iproto_type_name(req->type);
+	static char prefix[128];
+
+	snprintf(prefix, sizeof(prefix), "RAFT: rejecting %s (%d) "
+		 "request from origin_id %u replica_id %u term %llu",
+		 type_name ? type_name : "UNKNOWN", req->type,
+		 req->origin_id, req->replica_id,
+		 (long long)req->term);
+
+	return prefix;
+}
```

Comment:
Sure. I missed that. Thanks!
```diff
diff --git a/src/box/txn_limbo.c b/src/box/txn_limbo.c
index fce9d74ba..9c08cb652 100644
--- a/src/box/txn_limbo.c
+++ b/src/box/txn_limbo.c
@@ -35,6 +35,7 @@
 #include "journal.h"
 #include "box.h"
 #include "raft.h"
+#include "tt_static.h"
 
 struct txn_limbo txn_limbo;
 
@@ -813,19 +814,16 @@ txn_limbo_wait_persisted(struct txn_limbo *limbo)
  * Fill the reject reason with request data.
  * The function is not reenterable, use with care.
  */
-static char *
+static const char *
 reject_str(const struct synchro_request *req)
 {
        const char *type_name = iproto_type_name(req->type);
-       static char prefix[128];
 
-       snprintf(prefix, sizeof(prefix), "RAFT: rejecting %s (%d) "
-                "request from origin_id %u replica_id %u term %llu",
-                type_name ? type_name : "UNKNOWN", req->type,
-                req->origin_id, req->replica_id,
-                (long long)req->term);
-
-       return prefix;
+       return tt_sprintf("RAFT: rejecting %s (%d) request from origin_id %u "
+                         "replica_id %u term %llu",
+                         type_name ? type_name : "UNKNOWN", req->type,
+                         req->origin_id, req->replica_id,
+                         (long long)req->term);
 }
 
 /**
```

Reply:
I've been considering `tt_sprintf` in first place then dropped this idea because I wanted the reject string to be independent of any other helpers. Though I'm fine with it.